core/sql/optimizer/costmethod.cpp

```diff
@@ -13997,15 +13997,188 @@ CostMethodFastExtract::computeOperatorCostInternal(RelExpr* op,
+  countOfAsynchronousStreams_ = activePartitions_;
+} // CostMethodHbaseInsert::cacheParameters()
+
+
+
+// -----------------------------------------------------------------------
+// CostMethodHbaseInsert::computeOperatorCostInternal()
+// -----------------------------------------------------------------------
+Cost* CostMethodHbaseInsert::computeOperatorCostInternal(RelExpr* op,
+  const Context* myContext,
+  Lng32& countOfStreams)
+{
+  cacheParameters(op, myContext);
+  estimateDegreeOfParallelism();
+
+  // ------------------------------------------------------
+  // Save off our current estimated degree of parallelism.
+  // in the 'out' parameter; we might revise it below
+  // ------------------------------------------------------
+  countOfStreams = countOfStreams_;
+
+  CostScalar currentCpus =
+    (CostScalar)myContext->getPlan()->getPhysicalProperty()->getCurrentCountOfCPUs();
+  activeCpus_ = MINOF(countOfAsynchronousStreams_, currentCpus);
+
+  // update count of streams; the caller of the method uses this value
+  if ((countOfAsynchronousStreams_ > 0) &&
+      (countOfAsynchronousStreams_ < countOfStreams)
+      )
+    countOfStreams = (Lng32)countOfAsynchronousStreams_.getValue();
+
+
+  streamsPerCpu_ =
+    (countOfAsynchronousStreams_ / activeCpus_).getCeiling();
+
+  CostScalar noOfProbesPerStream(csOne);
+
+  // Determine the number of probes per stream. Use this number as
+  // the number of rows to insert (this is "per-stream" costing).
+
+  noOfProbesPerStream =
+    (noOfProbes_ / countOfAsynchronousStreams_).minCsOne();
+
+  // ************************************************************
+  // Compute the write/read cost for the insert
+  //
+  // ************************************************************
+
+  // ---------------------------------------------------------------------
+  // Synthesize the cost vectors.
+  // ---------------------------------------------------------------------
+  SimpleCostVector cvFR;
+  SimpleCostVector cvLR;
+
+  // For now, we don't bother to estimate CPU time, I/O time, transfer 
+  // time or idle time, since we really are only supporting the new 
+  // cost model.
+  //
+  // Future possible improvements:
+  //
+  // 1. Take into account HBase memstore insertion cost. The memstore
+  // uses a Red-Black tree which has o(n * log(n)) insertion cost. To
+  // model this correctly, we'd need to take into account the number
+  // of HBase regions rather than the number of ESPs, that is, to
+  // divide the number of probes by the number of HBase regions to find
+  // n. This cost will be paid no matter how many inserting streams
+  // there are so by itself this may not be interesting. It would only
+  // be interesting if there were a choice in the plan between inserting
+  // and not inserting (e.g. if we were considering bypassing the
+  // memstore, or if we were considering storing an intermediate result,
+  // neither of which are choices we examine today).
```

Comment:
I agree completely; this is what I was trying to say in the comments. Thanks for the tip on NATable::getRegionsBeginKey(). I'll add that to the comments on the next check-in.